### PR TITLE
ON HOLD: Trial testing system: take 2

### PR DESCRIPTION
**Pull request by ichorid**
It is all cleaned up now (continued from https://github.com/Tribler/py-ipv8/pull/245). Only commits relevant to Trial migration here.

### Test output
Basically nose still run the tests, nothing changed on the outside. However, inside the `unittest` object is provided by `twisted.trial` module. 

### Alternative reactor Nose runner
To ensure that the runtime is not increased due to https://github.com/Tribler/py-ipv8/pull/245#issuecomment-418434274 we have to use our own runner for `nosetest` utility to be able to select an alternative reactor. We only do this on Unix systems, though. By default we try to use `glib2reactor` that is stable, does not produce any spurious messages and does not hit https://github.com/Tribler/py-ipv8/pull/245#issuecomment-418434274 problem. If it not available, we fall back to `selectreactor` that is available for all platforms.

### Trial-based runner
In addition, we now have the option to use Trial-based runner, that automatically scans the dirs for tests. It does not collect the logs, though. However, in parallel mode, it runs the whole suite in under 1 second, so it is useful for doing quick checks (only on Unix platforms).
To use it, one must first regenerate the Twisted plugins cache with provided utility (or just run it more than one time).

### CODE REVIEW

File: ipv8/test/messaging/anonymization/test_hiddenservices.py [hunk 1]
```diff
@@ -28,6 +27,10 @@ def announce(self, info_hash):
             global_dht_services[info_hash] = [self.address]
 
 
+class Twisted(object):
```
Comment (qstokkink) [inline ipv8/test/messaging/anonymization/test_hiddenservices.py]: Why do we need this?

File: regen_plugins_cache.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+from twisted.plugin import IPlugin, getPlugins
```
Comment (qstokkink) [inline regen_plugins_cache.py]: I take it this is to fix the weird plugin loading thing. Please add comments here so nobody removes it in the future. Also, why not add this to the `__init__.py`?

File: run_nose.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+if __name__ == '__main__':
+    try:
+        from twisted.internet import glib2reactor as reactor
+    except ImportError:
+        from twisted.internet import selectreactor as reactor
+
+    reactor.install()
+
+    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
```
Comment (qstokkink) [inline run_nose.py]: Why is this here? Please add some comments.

**Comment by qstokkink** [comment]
Apparently Linux and Windows are running the old test scripts and Mac is running the new one, we need to look at this.

Also, when using `trial` Mac is showing an error. The unit tests for Mac still pass though. For some reason the uncaught exceptions are not being inspected properly.

**Comment by qstokkink** [comment]
We may want to first make sure this PR is merged and passing using the old test scripts, before we start using trial though.

File: ipv8/test/messaging/anonymization/test_hiddenservices.py (hunk 1 above)
Comment (ichorid) [reply ipv8/test/messaging/anonymization/test_hiddenservices.py]: the reason for this is too much alt+enter smashing in PyCharm...
Funny, though, that pylint don't catch it.

Comment (qstokkink) [reply ipv8/test/messaging/anonymization/test_hiddenservices.py]: We run a quick version of Pylint. The older version we run doesn't catch all of the errors, it does finish in 5 seconds instead of 2 minutes though.

**Comment by ichorid** [comment]
Also, please **DON'T** merge it until https://github.com/Tribler/tribler/pull/3842 is ready and approved for merging.

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
retest this please

File: ipv8/test/mocking/ipv8.py [hunk 4]
```diff
@@ -48,9 +48,9 @@ def __init__(self, crypto_curve, overlay_class, create_trustchain=False, create_
             self.endpoint.enable_community_statistics(self.overlay.get_prefix(), True)
 
     def unload(self):
-        self.endpoint.close()
+        self.overlay.unload()
```
Comment (devos50) [inline ipv8/test/mocking/ipv8.py]: What was the reason for swapping these statements? Since now it could be that there are incoming messages after unloading an overlay right?

File: run_all_tests_trial.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    config.parseOptions()
+    config['tbformat'] = 'verbose'
+    config['exitfirst'] = True
+
+    _initialDebugSetup(config)
+    trialRunner = _makeRunner(config)
+
+    test_loader = TestLoader()
+    test_loader.suiteClass = CustomSuite
+    test_suite = test_loader.discover('./ipv8/test', top_level_dir='.')
```
Comment (devos50) [inline run_all_tests_trial.py]: You could consider using `os.path.join` here (which is more compatible across OSses).

File: run_all_tests_unix.sh [hunk 6]
```diff
@@ -9,12 +9,7 @@ tty -s && tput bold
 echo -n "Starting IPv8 testsuite: "
 if nosetests --version >>/dev/null 2>&1; then
     echo "using test runner 'nosetests'!"
-    # Use nosetests2 if available and if not, try version possibly incompatible with python2
-    if [ -x "$(command -v nosetests2)" ]; then
-        test_command="nosetests2"
-    else
-        test_command="nosetests"
-    fi
+    test_command="python run_nose.py"
```
Comment (devos50) [inline run_all_tests_unix.sh]: It's not safe to assume that the `python` command one ones system always points to a Python 2 interpreter. Maybe we should check specifically for `python2` availability first (like we did with `nosetests2`)?

File: regen_plugins_cache.py (hunk 2 above)
Comment (ichorid) [reply regen_plugins_cache.py]: Included running this script in `run_all_tests_unix.py`

File: run_all_tests_unix.sh (hunk 6 above)
Comment (ichorid) [reply run_all_tests_unix.sh]: added a separate check for python interpreter version

File: run_nose.py (hunk 3 above)
Comment (ichorid) [reply run_nose.py]: umm, what's unclear?

Comment (ichorid) [reply run_nose.py]: Oh, that's the old comment.

File: run_all_tests_trial.py (hunk 5 above)
Comment (ichorid) [reply run_all_tests_trial.py]: did that

File: ipv8/test/mocking/ipv8.py (hunk 4 above)
Comment (ichorid) [reply ipv8/test/mocking/ipv8.py]: The reason was `None objects` kind of errors appearing in tests. At init, we first load the overlay, and then open the endpoint. Therefore, we *must* do it the other way around at shutdown, or the overlay tries to do stuff with the endpoint, which is not there...
The problem is, the overlay does not check if the endpoint object is still there.
For now its fine, no errors after this fix. But we'll have to clean up the init-shutdown sequences of overlays in the future.

**Comment by ichorid** [comment]
retest this please

File: ipv8/test/test_taskmanager.py [hunk 7]
```diff
@@ -34,22 +33,22 @@ def tearDown(self):
         self.tm.shutdown_task_manager()
         super(TestTaskManager, self).tearDown()
 
-    @twisted_wrapper
+    @inlineCallbacks
     @untwisted_wrapper
```
Comment (qstokkink) [inline ipv8/test/test_taskmanager.py]: Do we still need the `inlineCallbacks` and `untwisted_wrapper` with trial?

File: run_all_tests_trial.py (hunk 5 above)
Comment (qstokkink) [reply run_all_tests_trial.py]: `abspath` is not the same as `os.path.join`. `join` takes care of the OS dependent separator (`/`, which on Windows is `\`). The fix would be:

```python
os.path.join('.', 'ipv8', 'test')
```

File: ipv8/test/test_taskmanager.py (hunk 7 above)
Comment (ichorid) [reply ipv8/test/test_taskmanager.py]: I don't know.
Like, I really tried to minimize the invasiveness of the changes. The plan was to remove the cruft after this PR is merged and we already have working Trial-based tests.

Comment (qstokkink) [reply ipv8/test/test_taskmanager.py]: Fair enough.

**Comment by ichorid** [comment]
rebased on top of latest `master`.
@qstokkink, time to merge it!

**Comment by ichorid** [comment]
hallelujah! :angel:
